### PR TITLE
fix issue 368

### LINKED ISSUE #368: Occasional `bad use of unsafe.Pointer?` panics during GC
This panic has returned. Previously it was reported here:

https://github.com/segmentio/parquet-go/issues/309

We are seeing a similar error in a different component. It is rare, but sometimes happens in groups likely due to write or read patterns.

```
runtime: marked free object in span 0x7f73ee799028, elemsize=24 freeindex=239 (bad use of unsafe.Pointer? try -d=checkptr)
0xc146cd2000 alloc unmarked
0xc146cd2018 alloc unmarked
0xc146cd2030 alloc marked  
0xc146cd2048 alloc unmarked
0xc146cd2060 alloc unmarked
0xc146cd2078 alloc unmarked
0xc146cd2090 alloc unmarked
0xc146cd20a8 alloc unmarked
...
0xc146cd3668 free  marked   zombie
0x000000c146cd3668:  0x7265732e746f6f72  0x6d616e2e65636976 
0x000000c146cd3678:  0x00656e6364333065 
...
fatal error: found pointer to free object
```

Some examples:
image

### PR DESCRIPTION
**Pull request by achille-roussel**
Fixes #368 

Pulls the file from https://github.com/segmentio/parquet-go/pull/373

The _fix_ results in a performance regression on lookups into byte array dictionaries:
```
name                                    old time/op  new time/op  delta
Dictionary/Lookup/BYTE_ARRAY/N=100       106ns ± 1%   253ns ± 1%  +137.77%  (p=0.000 n=10+10)
Dictionary/Lookup/BYTE_ARRAY/N=1000      841ns ± 0%  2335ns ± 0%  +177.53%  (p=0.000 n=9+10)
Dictionary/Lookup/BYTE_ARRAY/N=10000    10.4µs ± 0%  25.0µs ± 0%  +141.28%  (p=0.000 n=9+10)
Dictionary/Lookup/BYTE_ARRAY/N=100000    188µs ± 4%   297µs ± 1%   +58.22%  (p=0.000 n=10+9)
Dictionary/Lookup/BYTE_ARRAY/N=1000000  3.59ms ±11%  4.47ms ± 3%   +24.72%  (p=0.000 n=9+10)

name                                    old value/s  new value/s  delta
Dictionary/Lookup/BYTE_ARRAY/N=100        940M ± 1%    395M ± 1%   -57.94%  (p=0.000 n=10+10)
Dictionary/Lookup/BYTE_ARRAY/N=1000      1.19G ± 0%   0.43G ± 0%   -63.97%  (p=0.000 n=9+10)
Dictionary/Lookup/BYTE_ARRAY/N=10000      963M ± 0%    399M ± 0%   -58.55%  (p=0.000 n=9+10)
Dictionary/Lookup/BYTE_ARRAY/N=100000     533M ± 4%    336M ± 1%   -36.82%  (p=0.000 n=10+9)
Dictionary/Lookup/BYTE_ARRAY/N=1000000    280M ±12%    224M ± 3%   -20.16%  (p=0.000 n=9+10)
```

### CODE REVIEW

File: row.go [hunk 1]
```diff
@@ -700,7 +700,7 @@ func reconstructFuncOfGroup(columnIndex int16, node Node) (int16, reconstructFun
 	return columnIndex, func(value reflect.Value, levels levels, row Row) (Row, error) {
 		var err error
 
-		if value.Kind() == reflect.Interface && value.IsNil() {
+		if value.Kind() == reflect.Interface {
```
Comment (achille-roussel) [inline row.go:703]: This was a fix for a separate issue which triggered in `TestIssue368`:

```
--- FAIL: TestIssue368 (0.01s)
panic: reflect: call of reflect.Value.MapIndex on interface Value [recovered]
	panic: reflect: call of reflect.Value.MapIndex on interface Value
```

**Review by joe-elliott — APPROVED** [review]
Using this branch I can no longer repro this issue. Nice work!

File: dictionary_amd64.go [hunk 2]
```diff
@@ -69,12 +71,26 @@ func (d *doubleDictionary) lookup(indexes []int32, rows sparse.Array) {
 
 func (d *byteArrayDictionary) lookupString(indexes []int32, rows sparse.Array) {
 	checkLookupIndexBounds(indexes, rows)
-	dictionaryLookupByteArrayString(d.offsets, d.values, indexes, rows).check()
+	// TODO: this optimization is disabled for now because it appears to race
```
Comment (joe-elliott) [inline dictionary_amd64.go:74]: Can we add the details of the CPU type that trigger this issue? Either in these comments or in the linked issue below?

That way if someone attempts a fix on this they can make sure to be testing on the CPU type that has a chance of reproing?

**Review by abraithwaite — APPROVED** [review]
Sucks about the performance regression, but at least it's only for a particular type!

LGTM!

**Pull request merged**